scripts/count_monty_params.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2025 Thousand Brains Project
+#
+# Copyright may exist in Contributors' modifications
+# and/or contributions to the work.
+#
+# Use of this source code is governed by the MIT
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+"""Count the number of parameters in the `dist_agent_1lm`  Monty model.
+
+This script counts the number of parameters in a Monty model, specifically the
+`dist_agent_1lm` pretrained model. This produces the result (4 million parameters)
+referenced in the Continual Learning section of the paper.
+
+The Monty model consists of a series of learned graphs for different objects. These
+are stored in a nested state dictionary, e.g.:
+
+    graph = state_dict["lm_dict"][lm_id]["graph_memory"]["mug"]["patch"]
+
+This script will count all the numerical values stored in these graphs, as a measure
+of the total number of parameters in the model.
+
+For clarity, we begin by breaking down the structure of the state dictionary.
+
+A Monty model's state_dict contains the following keys:
+- 'lm_dict'
+- 'sm_dict'
+- 'motor_system_dict'
+- 'lm_to_lm_matrix'
+- 'lm_to_lm_vote_matrix'
+- 'sm_to_lm_matrix'
+
+All of the information we are concerned with is in lm_dict. sm_dict and
+motor_system_dict contain information for logging and visualization purposes. The
+connectivity information in lm_to_lm_matrix, lm_to_lm_vote_matrix, and sm_to_lm_matrix
+is a negligible fraction of data, particularly in single-LM models.
+
+Within lm_dict, lm_id is the index of the Learning Module(s), of which only 0 is
+relevant for single-LM models. lm_dict does not contain any other keys.
+
+state_dict["lm_dict"][lm_id] contains:
+- 'graph_memory'
+- 'target_to_graph_id'
+- 'graph_id_to_target'
+The latter two are mappings between the object name and the index of the graph, which
+are needed to assess the performance of the model. They are of negligible size.
+
+state_dict["lm_dict"][lm_id]["graph_memory"] contains keys corresponding
+to all of the learned objects (e.g. 77 in the case of YCB).
+
+Each graph key (e.g. state_dict["lm_dict"][lm_id]["graph_memory"]["hammer"])
+is associated with a key corresponding to the sensory channel that was used to learn the
+object. This is typically "patch", but could be indexed (e.g. "patch_0", "patch_1") in
+multi-LM (hierarchical or voting) models.
+
+A graph can therefore be accessed with e.g.
+state_dict["lm_dict"][lm_id]["graph_memory"]["hammer"]["patch"].
+
+The accessed graph is a torch_geometric Data object, which has the following structure:
+
+Data(
+    x=[1188, 28],  # 28 features per node; not all of these are used in inference,
```

Comment:
How many of them do we use? Should we just store the features that we actually use?

Reply:
That's determined by what we specify tolerances for right? I guess just these two then:

```
            tolerances={
                "patch": {
                    "hsv": np.array([0.1, 0.2, 0.2]),
                    "principal_curvatures_log": np.ones(2),
                }
            },
```

From:
https://github.com/thousandbrainsproject/tbp.tbs_sensorimotor_intelligence/blob/58f37a15ff5857858e92fef3d1fb50ccb1011bcc/monty/configs/common.py#L103

Reply:
ok so that would be 3+2+9=14 out of 28 that we use. (9 for the pose features) That would cut down the overall number of model parameters quite a bit, right? Seems reasonable to account for that.

Reply:
Yeah although I guess we also have to use the `pose_fully_defined`, `node_ids` etc. I might just leave it for now if it's ok but yeah we can update it if we want to emphasize the point even more (it's already a difference of 4M parameters for our model vs. 86 million for the smallest ViT).